### PR TITLE
[19.09] make toolshed repo index understand yaml

### PR DESCRIPTION
**Pull request by martenson**
plus set up a cron-like service for indexing shed database using builtin uwsgi module

fixes https://github.com/galaxyproject/galaxy/issues/8880
also lazy implementation of https://github.com/galaxyproject/galaxy/issues/8881

cc @natefoo @mvdbeek

### CODE REVIEW

File: scripts/get_uwsgi_args.py [hunk 1]
```diff
@@ -99,6 +99,7 @@ def _get_uwsgi_args(cliargs, kwargs):
         'hook-master-start': ('unix_signal:2 gracefully_kill_them_all',
                               'unix_signal:15 gracefully_kill_them_all'),
         'py-call-osafterfork': True,
+        'cron': '0 -1 -1 -1 -1 python scripts/tool_shed/build_ts_whoosh_index.py -c config/tool_shed.yml --config-section tool_shed',
```
Comment (natefoo) [inline scripts/get_uwsgi_args.py]: Technically this will fail if you don't have a `config/tool_shed.yml`. I'm not super worried about that though.

**Review by natefoo — APPROVED** [review]

File: scripts/get_uwsgi_args.py (hunk 1 above)
Comment (natefoo) [reply scripts/get_uwsgi_args.py]: Actually... you can template `config_file` in to this to fix it.

**Comment by martenson** [comment]
When shut down uwsgi takes an extra minute to stop. Maybe it is waiting for the cron worker?

```
Toolbox index finished (103.400 ms)
[uwsgi-cron] command "python scripts/tool_shed/build_ts_whoosh_index.py -c config/tool_shed.yml --config-section tool_shed -d" running with pid 52708 exited after 4 second(s)
^CMon Nov 18 10:50:13 2019 - graceful shutdown triggered...
Gracefully killing worker 1 (pid: 50787)...
gateway "uWSGI http 1" has been buried (pid: 50788)
Mon Nov 18 10:51:14 2019 - worker 1 (pid: 50787) is taking too much time to die...NO MERCY !!!
worker 1 buried after 1 seconds
goodbye to uWSGI.
```

**Comment by martenson** [comment]
So the issue here is that cron runs on master uwsgi process, which we disable by default so uwsgi can be killed gracefuly on macos, however when cron is enabled uwsgi will automatically create the master process, hence the problem with shutting down on macos.

File: scripts/tool_shed/build_ts_whoosh_index.py [hunk 2]
```diff
@@ -98,18 +125,18 @@ def build_index(sa_session, whoosh_index_dir, path_to_repositories, hgweb_config
...
-        print(repos_indexed, 'repos (', repo.get('id'), ')')
+        # log.debug(str(repos_indexed) + ' repos (' + repo.get('name') + ')')
 
     tool_index_writer.commit()
     repo_index_writer.commit()
 
-    print("TOTAL repos indexed: ", repos_indexed)
-    print("TOTAL tools indexed: ", tools_indexed)
+    log.debug("Indexed repos: %s, tools: %s", repos_indexed, tools_indexed)
+    log.debug("Toolbox index finished %s", execution_timer)
```
Comment (natefoo) [inline scripts/tool_shed/build_ts_whoosh_index.py]: Maybe better to make these `info` level?

File: scripts/tool_shed/build_ts_whoosh_index.py [hunk 3]
```diff
@@ -2,52 +2,81 @@
...
 
 if sys.version_info > (3,):
     long = int
 
-logging.basicConfig(level='DEBUG')
-
-
-def build_index(sa_session, whoosh_index_dir, path_to_repositories, hgweb_config_dir):
+log = logging.getLogger()
+log.setLevel(logging.DEBUG)
```
Comment (natefoo) [inline scripts/tool_shed/build_ts_whoosh_index.py]: Should this be conditional on `-d/--debug`?

File: scripts/tool_shed/build_ts_whoosh_index.py (hunk 2 above)
Comment (martenson) [reply scripts/tool_shed/build_ts_whoosh_index.py]: Done, thanks!

File: scripts/tool_shed/build_ts_whoosh_index.py (hunk 3 above)
Comment (martenson) [reply scripts/tool_shed/build_ts_whoosh_index.py]: Done, thanks!

File: scripts/tool_shed/build_ts_whoosh_index.py [hunk 4]
```diff
@@ -98,18 +125,18 @@ def build_index(sa_session, whoosh_index_dir, path_to_repositories, hgweb_config
                                            repo_name=unicodify(repo.get('name')),
                                            repo_id=repo.get('id'))
             tools_indexed += 1
-            print(tools_indexed, 'tools (', tool.get('id'), ')')
+            # log.debug(str(tools_indexed) + 'tools (' + tool.get('id') + ')')
```
Comment (mvdbeek) [inline scripts/tool_shed/build_ts_whoosh_index.py]:
```suggestion
```

File: scripts/tool_shed/build_ts_whoosh_index.py [hunk 5]
```diff
@@ -98,18 +125,18 @@ def build_index(sa_session, whoosh_index_dir, path_to_repositories, hgweb_config
...
 
         repos_indexed += 1
-        print(repos_indexed, 'repos (', repo.get('id'), ')')
+        # log.debug(str(repos_indexed) + ' repos (' + repo.get('name') + ')')
```
Comment (mvdbeek) [inline scripts/tool_shed/build_ts_whoosh_index.py]:
```suggestion
```

File: scripts/tool_shed/build_ts_whoosh_index.py [hunk 6]
```diff
@@ -2,52 +2,81 @@
 Build indexes for searching the Tool Shed.
 Run this script from the root folder, example:
 
-$ python scripts/tool_shed/build_ts_whoosh_index.py -c config/tool_shed.ini --hgweb var/hgweb_config_dir/
+$ python scripts/tool_shed/build_ts_whoosh_index.py -c config/tool_shed.yml
 
-Make sure you adjusted your config to:
- * turn on searching via toolshed_search_on
- * specify whoosh_index_dir where the indexes will be placed
+Make sure you adjusted your Toolshed config to:
```
Comment (mvdbeek) [inline scripts/tool_shed/build_ts_whoosh_index.py]:
```suggestion
Make sure you adjust your Toolshed config to:
```

**Review by mvdbeek — APPROVED** [review]

**Pull request merged**